apis/Google.Cloud.Storage.V1/Google.Cloud.Storage.V1/UrlSigner.HmacBlobSigner.cs

```diff
@@ -0,0 +1,89 @@
+// Copyright 2023 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License").
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at 
+//
+// https://www.apache.org/licenses/LICENSE-2.0 
+//
+// Unless required by applicable law or agreed to in writing, software 
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and 
+// limitations under the License.
+
+using Google.Apis.Util;
+using System;
+using System.Globalization;
+using System.Security.Cryptography;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Google.Cloud.Storage.V1;
+
+public sealed partial class UrlSigner
+{
+    /// <summary>
+    /// Blob signer to implement signing using an HMAC secret.
+    /// </summary>
+    public sealed class HmacBlobSigner : IBlobSigner
+    {
+        private const string DefaultPrefix = "GOOG4";
+
+        private readonly string _secret;
+        private readonly string _prefix;
+
+        /// <summary>
+        /// Creates a new HMAC Blob Signer from an HMAC Key ID and Secret.
+        /// </summary>
+        /// <param name="keyId">The HMAC key ID. Must not be null or empty.</param>
+        /// <param name="secret">The HMAC key secret. Must not be null or empty.</param>
+        private HmacBlobSigner(string keyId, string secret) =>
+            (Id, _secret, _prefix) = (keyId.ThrowIfNullOrEmpty(nameof(keyId)), secret.ThrowIfNullOrEmpty(nameof(secret)), DefaultPrefix);
```

Comment:
Presumably the plan is that we can allow prefix customization at a later date if necessary...

Reply:
Yes, exactly. The prefix itself could have been harcoded at the moment, but the constructor is private and the factory method is there precisely in case we need to allow other prefixes in the future.